Add no-access table-level and schema-level permission

escherize opened the pull request:
This allows setting the `:blocked` data-view permission on the table and schema levels.

Prior to this we were only allowing this to be set on the database level.

- [x] allow `:blocked` to be saved for table level perms
- [x] ensure we are correctly handling `:blocked` as a value in the permission graph response
- [x] add test for saving table-level `:blocked`
- [x] add test for permission graph response value for a table-level `:blocked` perm 
- [x] rename function for verification of data perms
- [x] write a test where you run a query with a user that has data-perms, but not create-query
    - [x] Need to make sure [this exhaustive test output](https://github.com/metabase/metabase/pull/46542/files#diff-4ea27560c9fa50886e19f48ebcca2e87f13e1d0d5b182ca8645fc0ad04b9aa8cR113-R122) is the correct expected behavior
- [x] Ensure native queries for databases with blocked tables cannot run


---
Unrelated fixes:

- [x] Handle listing tables for sqlite, with it's `""` schema.

escherize committed:
allow `:blocked` to be saved for table level perms

escherize committed:
Adds 2 tests for table level blocked permission settings

- N.B. these are NOT ENFORCED YET

johnswanson committed:
WIP: Pairing on making perms checking less wild

escherize committed:
update test that asserted we cannot set block on tables (we can)

escherize committed:
cleanup, update docs, and add a test for view-data perm only

- Added a test where we have data permissions, but not create query, and
  I think it is failing when we have create query and blocked data
  permissions.
- renamed some functions from check-x -> has-x? since they return a
  value instead of throwing now

escherize committed:
Revert "WIP: Pairing on making perms checking less wild"

Keep the same behavior, but stick with the saner flow control

This reverts commit 63bcb5b401138555a081378574b15d6506469e60.

escherize committed:
update docs

escherize committed:
update test to be passing

- TODO: make sure it's correct w.r.t. perm settings

escherize commented on enterprise/backend/test/metabase_enterprise/sandbox/api/card_test.clj:
```diff
@@ -102,6 +102,35 @@
+                 [:legacy-no-self-service :query-builder-and-native "Request Permitted"]
+                 [:legacy-no-self-service :query-builder            "Request Permitted"]
+                 [:legacy-no-self-service :no                       "You do not have permissions to run this query."]
+                 [:blocked                :query-builder-and-native "Request Permitted"]
+                 [:blocked                :query-builder            "Request Permitted"]
```
I think these are actually an impossible combo on the table level. db level too, but we are testing tables here.

escherize commented on enterprise/backend/test/metabase_enterprise/sandbox/api/card_test.clj:
```diff
@@ -102,6 +102,35 @@
+                 [:unrestricted           :query-builder            "Request Permitted"]
+                 [:unrestricted           :no                       "Request Permitted"]
+                 [:legacy-no-self-service :query-builder-and-native "Request Permitted"]
+                 [:legacy-no-self-service :query-builder            "Request Permitted"]
```
I think these are actually an impossible combo on the table level.

escherize committed:
Allow schema level blocked setting in permgraph

escherize committed:
remove invalid test cases

- continue to have a forcing function to test newly added perms

escherize committed:
conform function output

escherize committed:
ensure a single blocked table blocks native queries to its DB

escherize committed:
update error message

- we now catch this error in `metabase.models.query.permissions/has-perm-for-query?`

escherize committed:
we now check for data permissions to process query for card

escherize commented:
I think I found a bug while fixing tests on this branch.

We were letting people run queries for which they had missing data perms. That seems wrong to me, so I updated the test to show that you need data perms for the queries you are trying to run [in this commit](https://github.com/metabase/metabase/pull/46542/commits/008c90bf266f6b233d43fd65f26bee354bbf6504).

escherize committed:
add more explanation to what we are testing

- to help see why it fails on CI and passes locally

johnswanson commented:
So I was looking into the bug you mentioned, and it seems like the test is brittle, and *doesn't* pass on `master` when it's working correctly. The problem is that in a fresh REPL, running that test results in data perms like the following:

```
| :group_id | :db_id | :table_id |                   :perm_type |               :perm_value |
|-----------+--------+-----------+------------------------------+---------------------------|
|         1 |      2 |           | :perms/manage-table-metadata |                       :no |
|         1 |      2 |           |       :perms/manage-database |                       :no |
|         1 |      2 |           |        :perms/create-queries | :query-builder-and-native |
|         1 |      2 |           |             :perms/view-data |             :unrestricted |
|         1 |      2 |           |      :perms/download-results |         :one-million-rows |
```

But why - since we're using `mt/with-no-data-perms-for-all-users!`? Because... at the point we use `mt/with-no-data-perms-for-all-users!` the database doesn't exist yet. Then, when we create it, we give it the default permissions - `:unrestricted`. So when the test runs *for the first time* (when the test DB doesn't exist yet) it passes. I think this is why it's been passing in CI.

But in reality, it shouldn't - and the second time I run it, it fails with an error that the user is blocked. So that's good - the bug is in the test, but the actual permissions behavior is like we want.

I was able to get consistent (failing) behavior by moving the `mt/with-no-data-perms-for-all-users!` wrapper down, just after the `mt/with-temp` block. Then the test consistently fails with

```
expected: (= [[1] [2]] (mt/rows (process-query-for-card child-card)))
   error: clojure.lang.ExceptionInfo: Blocked: you are not allowed to run queries against Database 2.
```

johnswanson commented on src/metabase/query_processor/middleware/permissions.clj:
```diff
@@ -90,8 +90,9 @@
+            (throw (query-perms/perms-exception required-perms))))
 
         ;; set when querying for field values of dashboard filters, which only require
         ;; collection perms for the dashboard and not ad-hoc query perms
```
Can't leave the comment on the right line here, but isn't the below check under `*param-values-query*` doing the same thing as you're changing on L93-95? Should we change it to match?

escherize replied:
Looks like it. I've updated it to match the more straightforward style.

escherize committed:
remove excess `def`

escherize committed:
Add test for table-level data X collection perms

- update test found to be in-error

escherize committed:
update param values qp permission check style

escherize committed:
set view-data and create-query explicitly

escherize committed:
set viewdata and createquery explicitly in qp test

johnswanson commented on test/metabase/api/card_test.clj:
```diff
@@ -3658,8 +3723,27 @@
+            (testing "create-queries = query-builder-and-native is sufficent to run the query"
+              (mt/with-restored-data-perms!
+                (data-perms/set-database-permission! (perms-group/all-users) (mt/id) :perms/view-data :blocked)
+                (data-perms/set-database-permission! (perms-group/all-users) (mt/id) :perms/create-queries :query-builder-and-native)
```
I was a little confused by this - if you look at the code for `set-database-permission!`, it will automatically set `:perms/view-data` to `:unrestricted` if you set `:perms/create-queries` to anything but `:no`. So i think what's actually going on here is that you need `:unrestricted` data access, and your `:perms/create-queries` permission doesn't matter.

escherize replied:
Fixed

johnswanson commented on test/metabase/api/card_test.clj:
```diff
@@ -3658,8 +3723,27 @@
                 (is (mi/can-read? child-card)))
-              (is (= [[1] [2]]
-                     (mt/rows (process-query-for-card child-card)))))))))))
+              (is (= :blocked
```
This test was a little confusing for me because the top-level `testing` block says "Should be able to run a Card with another Card as its source query with just perms for the former" but then we end up testing a lot of different scenarios (perms for neither, then different perms options for one).

escherize replied:
Thanks for pointing that out, fixed

johnswanson reviewed (changes requested):
I think you need a test like the following:

```
(deftest cannot-run-any-native-queries-when-blocked-test
  (mt/with-non-admin-groups-no-root-collection-perms
    (mt/with-temp [:model/Collection allowed-collection    {}
                   :model/Collection disallowed-collection {}
                   :model/Card       parent-card           {:dataset_query {:database (mt/id)
                                                                            :type     :native
                                                                            :native   {:query "SELECT id FROM venues ORDER BY id ASC LIMIT 2;"}}
                                                            :database_id   (mt/id)
                                                            :collection_id (u/the-id disallowed-collection)}
                   :model/Card       child-card            {:dataset_query {:database (mt/id)
                                                                            :type     :query
                                                                            :query    {:source-table (format "card__%d" (u/the-id parent-card))}}
                                                            :collection_id (u/the-id allowed-collection)}]
      (mt/with-no-data-perms-for-all-users!
        (data-perms/set-database-permission! (perms-group/all-users) (mt/id) :perms/view-data :blocked)
        (data-perms/set-table-permission! (perms-group/all-users) (mt/id :venues) :perms/view-data :unrestricted)
        (perms/grant-collection-read-permissions! (perms-group/all-users) allowed-collection)
        (clojure.pprint/print-table (t2/select :model/DataPermissions :db_id (mt/id)))
        (letfn [(process-query-for-card [card]
                  (mt/user-http-request :rasta :post (format "card/%d/query" (u/the-id card))))]
          (is (thrown-with-msg?
               clojure.lang.ExceptionInfo
               #"You do not have permissions to run this query."
               (mt/rows (process-query-for-card child-card)))
              "Someone with `:blocked` permissions on ANY table in the database cannot run ANY card with native queries, including as a source for another card."))))))
```

Playing around locally, I added the above and it failed - we ran the query successfully. I think it actually might be a problem with `set-table-permission!` though, because when I looked at the data permissions they were all `:unrestricted`, which isn't right 😅 

I'd look in `set-table-permission!` - I think the problem might be this bit here:

```
                                            :perm_value  (case existing-db-perm-value
                                                           ;; If the previous database-level permission can't be set at
                                                           ;; the table-level, we need to provide a new default
                                                           :query-builder-and-native :query-builder
                                                           :blocked                  :unrestricted
                                                           existing-db-perm-value)
```

`:blocked` _can_ now be set at the table level, so we should leave the existing rows alone.

escherize committed:
Respond to review comments (which fixes a case)

escherize committed:
Merge branch 'master' into add-no-access-table-level-permission

johnswanson reviewed (approved):
Nice, this looks great!

escherize committed:
setting a table to blocked: leave other tables the same

sloansparger committed:
[Permissions] Add "No access" schema/table permission (#46509)

* first pass

* refactors downgrading native permission logic and updates calculation so that "No access" downgrades native permissions to "No"

* stub for permissions help info on table block

* modal changes wip, updates downgrading create queries permissions to all happen at a single call site

* clean up, sandboxing modal copy changes, removes rekoke/limit access modal changes to make the diff smaller and move code to a seperate PR

* updates permissions help section to contain the final copy

* sandboxing copy fix and remove modal that was dropped from requirements

* adds blocked at the schema level, updates no access copy to blocked, updates permissions help section to contain new blocked and schema level changes

* fixes failed unit and e2e tests after sandboxing copy changes

* improve the block e2e test to include table blocking

* fixes failing blocked test, fixes other schemas create queries permissions getting correct with one schema was droped to blocked view data access, fixes a bug that prevents the save bar from going away when all permissions for group are set to the default values

* clean up

* remove color changes

* prevents parent being set to blocked preventing edits for children entities

* add new hasPermissionValueInSubgraph fn, adds modal to warn users we have to upgrade the view data permissions when they upgrade create queries permissions when a child entity is set to blocked

* adds test coverage for new modal

* removes unused function, adds new updateEntityPermission fn to help consolidate some logic elsewhere

* unit test fix and type fix

* most pr feedback

* updates the confirmation modal copy when changing a parent entity that contains a child with blocked permissions and/or sandboxed children, adds test coverage for that, adds test coverage for permission view data column not appearing in oss

* type fix

sloansparger committed:
[Permissions] Add e2e test coverage for blocked permissions enforcements (#46663)

* adds test coverage for enforcement of blocked permissions

* moves tests around based on pr feedback

imrkd commented:
/perf

sloansparger committed:
copy changes

sloansparger committed:
Merge branch 'master' into add-no-access-table-level-permission

sloansparger committed:
adds fix to make sure that blocked permissions are not removed from sibling tables that have the create queries permissions upgraded (#46854)

escherize committed:
Fix table name lookup for dbs w/ 1 schema per db

escherize committed:
add test for blank schema identifiers

escherize committed:
Refine sandboxed user perms for query builder access (#46939)

* Refine sandboxed user perms for query builder access

- Limit create-queries permissions to unblocked tables only
- Check user permissions for each table before granting query builder access
- Prevent querying of blocked joined tables from query builder for sandboxed users

* Adjust permissions for sandboxed users

- Grant view-data permissions only for unblocked tables
- Revert create-queries permissions to all tables in sandbox
- Remove unnecessary intermediate variable

* when sandboxing we no longer grant unrestricted view perms for blocked tables

escherize committed:
Merge branch 'master' into add-no-access-table-level-permission

escherize committed:
Merge branch 'master' into add-no-access-table-level-permission

escherize commented:
This last merge added [Include missing opts arg for transform->nested](https://github.com/metabase/metabase/pull/46944), which fixes some unrelated CI failures.

noahmoss commented on enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj:
```diff
@@ -259,10 +260,24 @@
-      {:perms/view-data (zipmap table-ids (repeat :unrestricted))
+    (let [table-ids (sandbox->table-ids sandbox)
+          table-id->db-id (if (seq table-ids)
+                            (m/index-by :id (t2/select :model/Table :id [:in table-ids]))
```
The DB call here makes me a little nervous since we've had recent perf issues in sandboxing code. Two ideas:
* Use `metabase.models.database/table-id->database-id` which is memoized per app DB. Issue there is that if the cache isn't populated it'll do a separate DB call for each table. Maybe we could add a batched version that does a single DB call for anything not in the cache but that sounds more complex.
* Update `metabase-enterprise.sandbox.api.util/enforced-sandboxes-for-user` to fetch DB IDs via a join when populating the sandbox cache for a given request. I like this idea better.

escherize replied:
I am working on implementing the second option here. I noticed since we already hydrate tables in the query for enforced-sandboxes-for-users, there's no need to join (we already get the table-id <-> db-id info. 

But! I noticed something interesting: I have sandboxed orders (see screenshot), and when I go to save a vanilla query on orders using a nomral user, we call `sandbox->required-perms`, where the sandbox and `@data-perms/*sandboxes-for-user*` have different sets of table ids.

```clojure
;; sandbox has:
{3 1, 5 1}

;; @data-perms/*sandboxes-for-user* has:
(map
 (comp (juxt :id :db_id) :table)
 (@data-perms/*sandboxes-for-user*)) => ([5 1])
```


![Screenshot](https://github.com/user-attachments/assets/2cb4b18e-4e6a-4a5f-ad39-cdb4321150cf)



This doesn't seem to find the People table:
``` 
;; sandbox has:
{3 1, 5 1}

;; @data-perms/*sandboxes-for-user* has
(map
 (comp (juxt :id :db_id) :table)
 (@data-perms/*sandboxes-for-user*)) => ([5 1])
```

Since this seems to not be a 1:1 drop in, I will start writing option 1. Curious to hear your thoughts

noahmoss commented on enterprise/backend/test/metabase_enterprise/advanced_permissions/models/permissions/block_permissions_test.clj:
```diff
@@ -119,12 +119,12 @@
           (is (not (t2/exists? GroupTableAccessPolicy :group_id group-id))))))))
 
 (deftest update-graph-data-perms-should-delete-block-perms-test
- (testing "granting data permissions for a table should delete existing block permissions"
+ (testing "granting data permissions for a table should not delete existing block permissions"
```
nit - update test name?

escherize replied:
Done

noahmoss commented on enterprise/backend/test/metabase_enterprise/sandbox/api/card_test.clj:
```diff
@@ -102,6 +102,37 @@
+             (- (* (-> data-perms/Permissions :perms/view-data :values count)
+                   (-> data-perms/Permissions :perms/create-queries :values count))
+                (count invalid-cases)))
+          "Please test these permissions settings behaviors exhaustively: if you add perms, add the tests for them.")
```
💪

noahmoss commented on src/metabase/models/data_permissions.clj:
```diff
@@ -608,8 +611,7 @@
 
-  Block permissions (i.e. :perms/view-data :blocked) can only be set at the database-level, despite :perms/view-data
-  being a table-level permission."
+  Block permissions (i.e. :perms/view-data :blocked) can be set at the table or database-level."
```
I'd support just removing this line — previously it was calling out a special case, which is no longer special

escherize replied:
Done

escherize committed:
Update enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj

remove blank line

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

escherize committed:
- make coalesce-test exhaustive (except for sandbox)

escherize committed:
Update enterprise/backend/src/metabase_enterprise/sandbox/query_processor/middleware/row_level_restrictions.clj

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

escherize committed:
t2/select ... -> database/table-id->database-id

escherize committed:
Merge branch 'add-no-access-table-level-permission' of github.com:metabase/metabase into add-no-access-table-level-permission

escherize committed:
update comment

sloansparger committed:
[Permissions] Prevent "Granular" option in DB View Data options from changing permissions to unrestricted (#46976)

* fix

* adds back most of the code and limits it to only happen with impersonations, updates test to handle differing logic between the two flows

* removes test that is not longer needed

escherize committed:
more sandbox join table perms tests

escherize committed:
Merge branch 'add-no-access-table-level-permission' of github.com:metabase/metabase into add-no-access-table-level-permission

sloansparger committed:
Merge branch 'master' into add-no-access-table-level-permission
